### PR TITLE
CLI: Add command to generate annotations file

### PR DESCRIPTION
**Pull request by yannbf**
Issue: N/A

## What I did

The CLI now has a command to generate preset annotations file. This will get framework presets, addon presets and project annotation (preview.js) and write to a file that might look like this:

```
import * as frameworkPreset from "@storybook/react/dist/esm/client/docs/config";
import * as frameworkPreset2 from "@storybook/react/dist/esm/client/preview/config";
import * as addonPreset from "@storybook/addon-docs/dist/esm/frameworks/common/config";
import * as addonPreset2 from "@storybook/addon-actions/dist/esm/preset/addDecorator.js";
import * as addonPreset3 from "@storybook/addon-actions/dist/esm/preset/addArgs.js";
import * as addonPreset4 from "@storybook/addon-backgrounds/dist/esm/preset/addDecorator.js";
import * as addonPreset5 from "@storybook/addon-backgrounds/dist/esm/preset/addParameter.js";
import * as projectAnnotations from "./preview.ts";

export default [addonPreset, frameworkPreset, frameworkPreset2, addonPreset2, addonPreset3, addonPreset4, addonPreset5, projectAnnotations];
```

## How to test

1 - Build CLI
2 - Use it locally with `sb generate-annotations` in a Storybook project

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/cli/src/generate-annotations.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  return dedent`
+    ${presetImports.join('')}
+    export default [${presetNames.join(', ')}];
+  `;
+};
+
+export async function writeAnnotationsFile({ configDir = '.storybook', cjs = false }) {
+  try {
+    logger.info(`=> Generating preset annotations file`);
+    console.log({ configDir });
```
Comment (shilman) [inline lib/cli/src/generate-annotations.ts:55]: console.log?

**Review by shilman — COMMENTED** [review]
We should probably wait to merge this until https://github.com/storybookjs/storybook/pull/17755 is merged

File: lib/cli/src/generate-annotations.ts (hunk 1 above)
Comment (yannbf) [reply lib/cli/src/generate-annotations.ts]: oops

File: lib/cli/src/generate-annotations.test.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      const content = generateAnnotations.getAnnotationsFileContent([
+        '../node_modules/@storybook/addon-actions/dist/esm/preset/addDecorator.js',
+      ]);
+
+      expect(content).toEqual(dedent`
+        import * as addonPreset from "@storybook/addon-actions/dist/esm/preset/addDecorator.js";
+        
+        export default [addonPreset];
+      `);
+    });
```
Comment (tmeasday) [inline lib/cli/src/generate-annotations.test.ts:25]: What's going on with this test?

File: lib/cli/src/generate-annotations.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import { viewLayers } from './view-layers';
+
+export const getAnnotationsFileContent = (presets: string[], replaceEsmWithCjs?: boolean) => {
+  const presetNames: string[] = [];
+  const presetImports: string[] = [];
+
+  let previewFile;
+  const [firstImport] = presets.slice(-1);
+  if (firstImport.indexOf('preview') !== -1) {
```
Comment (tmeasday) [inline lib/cli/src/generate-annotations.ts:15]: This check isn't going to be good if we are calling the annotation files `.../preview.js`

File: lib/cli/src/generate-annotations.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    presetNames.push(finalName);
+    presetImports.push(`import * as ${finalName} from "${preset.path}";\n`);
+  });
+
+  if (previewFile) {
+    // we add ./ to relative paths because they are not provided by default
+    const previewPath = previewFile === 'preview' ? `./${previewFile}` : previewFile;
+
+    presetNames.push('projectAnnotations');
+    presetImports.push(`import * as projectAnnotations from "./${previewPath}";\n`);
```
Comment (tmeasday) [inline lib/cli/src/generate-annotations.ts:43]: Wouldn't this end up with `././`?

**Review by tmeasday — APPROVED** [review]
Generally good! Some nitpicks

**Review by shilman — COMMENTED** [review]
I'm unclear on the workflow here. Is the idea that the user will run this to generate an annotations file, and then will edit it down by hand to only include the pieces that they want? Based on our discussions I thought we concluded that typically users would build these files "constructively" by adding the minimal number of decorators etc. that they might need, rather than "destructively" by deleting. But maybe I misunderstood!

**Review by shilman — COMMENTED** [review]
(oops one more dupe... 🤷 )

**Review by shilman — COMMENTED** [review]
(oops duplicate)

**Review by shilman — COMMENTED** [review]
(oops another duplicate?)

File: lib/cli/src/generate-annotations.test.ts (hunk 2 above)
Comment (yannbf) [reply lib/cli/src/generate-annotations.test.ts]: I forgot to update it! 😅  thanks for the catch

File: lib/cli/src/generate-annotations.ts (hunk 4 above)
Comment (yannbf) [reply lib/cli/src/generate-annotations.ts]: Yup! great catch. fixed

File: lib/cli/src/generate-annotations.ts (hunk 3 above)
Comment (yannbf) [reply lib/cli/src/generate-annotations.ts]: Fixed!

**Comment by yannbf** [comment]
> 

Generally yes, but I thought we were going to get this PR in because it would make our experimentation easier?

cc @tmeasday

**Review by shilman — APPROVED** [review]
Works for me. I'm not enthusiastic about it, but I'm fine if it's convenient.

We can merge & re-address in 7.0 CLI refactor if needed

**Comment by yannbf** [comment]
@tmeasday after discussing more about this with @shilman we decided to keep this open until https://github.com/storybookjs/storybook/pull/17755 is resolved.

File: lib/core-common/src/presets.ts [hunk 5]
```diff
@@ -374,3 +375,19 @@ export function loadAllPresets(
...
+    configDir,
+  } as LoadPresetOptions);
+
+  const addonPresets = await presets.apply('previewAnnotations', [], { configDir });
+  const configs = [
+    // load addon presets
+    ...addonPresets,
+    // load preview.js
+    loadPreviewOrConfigFile({ configDir }),
+  ].map((configPath) => relative(configDir, configPath));
```
[the remark this answers is not in the document]
Comment (yannbf) [reply lib/core-common/src/presets.ts]: good call. done!

**Comment by yannbf** [comment]
Closing this as might be too impactful for the changes we are planning to do in 7.0. If we ever want this again we can revive the PR.

**Pull request closed without merging**